compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertToDma.cpp

```diff
@@ -146,129 +154,400 @@ LogicalResult unPackDmaInputs(IREE::LinalgExt::UnPackOp unPackOp,
+                                      SmallVector<OpFoldResult> &sizes,
+                                      SmallVector<OpFoldResult> &strides) {
+  auto reassociationIndices = collapseOp.getReassociationIndices();
+  ArrayRef<int64_t> inputShape = collapseOp.getSrcType().getShape();
+  ArrayRef<int64_t> resultShape = collapseOp.getType().getShape();
+  uint64_t resultRank = resultShape.size();
+  MLIRContext *ctx = collapseOp.getContext();
+
+  // Set strides to inner-most stride in each reassocation group.
+  for (auto reassociation : llvm::enumerate(reassociationIndices)) {
+    uint64_t index = reassociation.index();
+    uint64_t dim = reassociation.value().back();
+    strides[index] = strides[dim];
+  }
+  strides.resize(resultRank);
+
+  // Set sizes to output shape, and check that all dims are static.
+  sizes.clear();
+  for (int64_t dim : resultShape) {
+    if (dim == ShapedType::kDynamic) {
+      return collapseOp.emitOpError(
+          "has a dynamic shape which is currently unsupported.");
     }
-    strides = getAsIndexOpFoldResult(ctx, stridesI64);
-    operandOp = subviewOp.getSource().getDefiningOp();
-    sizes = subviewOp.getMixedSizes();
-    if (llvm::any_of(sizes, [](OpFoldResult fr) {
-          return !getConstantIntValue(fr).has_value();
-        })) {
-      return subviewOp->emitOpError(
-          " has dynamic shape that is not supported by the target dma op.");
+    sizes.push_back(getAsIndexOpFoldResult(ctx, dim));
+  }
+
+  // Offsets - merge reassocation groups.
+  SmallVector<OpFoldResult> collapsedOffsets;
+  for (auto reassociation : llvm::enumerate(reassociationIndices)) {
+    auto dims = reassociation.value();
+    SmallVector<int64_t> localStrides(dims.size(), 1);
+    for (uint64_t i = 1; i < dims.size(); ++i) {
+      uint64_t dim = dims.size() - i - 1;
+      localStrides[dim] = localStrides[dim + 1] * inputShape[dims[dim + 1]];
     }
+    OpBuilder builder(ctx);
+    builder.setInsertionPoint(collapseOp);
+    OpFoldResult combination = getLinearCombination(
+        builder, collapseOp.getLoc(),
+        ArrayRef<OpFoldResult>(offsets.begin() + dims[0],
+                               offsets.begin() + dims.back() + 1),
+        localStrides);
+    collapsedOffsets.push_back(combination);
+  }
+  offsets = collapsedOffsets;
+  assert(offsets.size() == sizes.size() && sizes.size() == strides.size() &&
+         "mismatch in the number of offsets, sizes and strides");
+  return success();
+}
+
+/// Update the offsets, sizes, and strides from an expand shape operation.
+LogicalResult updateFromExpandShape(memref::ExpandShapeOp expandShapeOp,
+                                    SmallVector<OpFoldResult> &offsets,
+                                    SmallVector<OpFoldResult> &sizes,
+                                    SmallVector<OpFoldResult> &strides) {
+  MLIRContext *ctx = expandShapeOp.getContext();
+  auto reassociationIndices = expandShapeOp.getReassociationIndices();
```

Comment:
Don't use `auto` here.

Reply:
I intentionally didn't use `auto` here, because IMO  `SmallVector<SmallVector<long, 2>, 4>` is uninformative. Why pollute code (and brain) with the detail that someone years ago thought that 4 groups of size 2 seemed like a good choice for static sizes?